graphix/open_graph.py

```diff
@@ -0,0 +1,383 @@
+        False
+        """
+        if not isinstance(other, Measurement):
+            return NotImplemented
+
+        return np.allclose(self.angle, other.angle) and self.plane == other.plane
+
+    def is_z_measurement(self) -> bool:
+        """Indicates whether it is a Z measurement
+
+        Example
+        -------
+        >>> from graphix.open_graph import Measurement
+        >>> Measurement(0.0, "XY").is_z_measurement()
+        True
+        >>> Measurement(0.0, "YZ").is_z_measurement()
+        False
+        >>> Measurement(0.1, "XY").is_z_measurement()
+        False
+        """
+        return np.allclose(self.angle, 0.0) and self.plane == "XY"
+
+
+class OpenGraph:
+    """Open graph contains the graph, measurement, and input and output
+    nodes. This is the graph we wish to implement deterministically
+
+    :param inside: the underlying graph state
+    :param measurements: a dictionary whose key is the ID of a node and the
+        value is the measurement at that node
+    :param inputs: a set of IDs of the nodes that are inputs to the graph
+    :param outputs: a set of IDs of the nodes that are outputs of the graph
+
+    Example
+    -------
+    >>> import networkx as nx
+    >>> from graphix.open_graph import OpenGraph, Measurement
+    >>>
+    >>> inside_graph = nx.Graph([(0, 1), (1, 2), (2, 0)])
+    >>>
+    >>> measurements = {i: Measurement(0.5 * i, "XY") for i in range(2)}
+    >>> inputs = [0]
+    >>> outputs = [2]
+    >>> og = OpenGraph(inside_graph, measurements, inputs, outputs)
+    """
+
+    # All the measurement, inputs and outputs information is stored in this
+    # graph. This is to simplify certain operations such as composition of
+    # graphs and checking equality.
+    # These attributes shouldn't be accessed directly by the user, instead,
+    # they should use the inputs(), outputs(), and measurements() properties.
+    #
+    # The following attributes are stored in the nodes:
+    #
+    # Non-output nodes:
+    # * "measurement" - The Measurement object for the node
+    #
+    # Input nodes:
+    # * "is_input"    - The value is always True
+    # * "input_order" - A zero-indexed integer used to preserve the ordering of
+    #                   the inputs
+    #
+    # Output nodes:
+    # * "is_output"    - The value is always True
+    # * "output_order" - A zero-indexed integer used to preserve the ordering of
+    #                    the outputs
+    _inside: nx.Graph
+
+    def __eq__(self, other):
+        """Checks the two open graphs are equal
+
+        This doesn't check they are equal up to an isomorphism"""
+
+        g1 = self.perform_z_deletions()
+        g2 = other.perform_z_deletions()
+
+        return (
+            g1.inputs == g2.inputs
+            and g1.outputs == g2.outputs
+            and nx.utils.graphs_equal(g1._inside, g2._inside)
+            and g1.measurements == g2.measurements
+        )
+
+    def __init__(
+        self,
+        inside: nx.Graph,
```

Comment:
what's the motivation to make this a class rather than functions for each directions of transformations, zx-> dataset (nx.graph, in, out, measurements) and (pattern or (nx.graph, in, out, measurements)) -> zx?

Reply:
One reason is that since these four items (nx.graph, in, out, measurements) are all related and would not often be used separately (at least I can't think of many situations), so I felt it was better to hold them together in a class since it represents one logical object. In our compiler, the Open graph has many more methods and we have functions that take the Open graph as an argument, so it makes sense for us to hold it all together as one object.

Graphix is a bit different than our compiler though. The purpose of putting the open graph definition in graphix is so that we can standardise our interfaces and develop interchangeable components. We could define the interface for open graphs to be the tuple of (nx.graph, in, out, measurement) to simplify things, I think this is probably okay for graphix's purposes at present. Though this way has the disadvantage of our interface being an implicit convention, something new developers may not realise when they start to use our softwares.

What do you think? Do you think we should keep the class or just use the tuple?

Reply:
Thanks for the clarification! 

> The purpose of putting the open graph definition in graphix is so that we can standardise our interfaces and develop interchangeable components.

this is indeed a good point, and generic open graph interface would help standardizing the interface. I think current implementation would do the job nicely.